File: project_euler/problem_601/solution.py
from math import gcd

def solve():
    # streak(n) = k: smallest k such that n+k NOT divisible by k+1
    # streak(n) >= s: n+j divisible by j+1 for j=1..s-1
    #   i.e., n ≡ 1 (mod lcm(2,3,...,s))
    # streak(n) = s: streak >= s but not >= s+1
    # P(s, N) = floor((N-2)/lcm(2,...,s)) - floor((N-2)/lcm(2,...,s+1))
    # Special case s=1: P(1,N) = (N-2) - floor((N-2)/2)

    def lcm(a, b):
        return a // gcd(a, b) * b

    # L[k] = lcm(2, 3, ..., k)
    L = {1: 1}
    cur = 1
    for k in range(2, 34):
        cur = lcm(cur, k)
        L[k] = cur

    def P(s, N):
        Nm2 = N - 2
        if s == 1:
            c1 = Nm2
        else:
            Ls = L[s]
            c1 = Nm2 // Ls if Ls <= Nm2 else 0
        Ls1 = L[s + 1]
        c2 = Nm2 // Ls1 if Ls1 <= Nm2 else 0
        return c1 - c2

    ans = sum(P(i, 4**i) for i in range(1, 32))
    print(ans)

solve()

File: project_euler/problem_601/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // streak(n) = k: smallest k s.t. n+k NOT div by k+1
    // streak(n) >= s: n ≡ 1 (mod lcm(2,...,s))
    // P(s,N) = floor((N-2)/lcm(2,...,s)) - floor((N-2)/lcm(2,...,s+1))
    // Answer = sum_{i=1}^{31} P(i, 4^i)

    typedef __int128 lll;

    auto my_gcd = [](lll a, lll b) -> lll {
        while (b) { a %= b; swap(a, b); }
        return a;
    };

    // L[k] = lcm(2, 3, ..., k)
    vector<lll> L(34, 1);
    lll cur = 1;
    for (int k = 2; k <= 33; k++) {
        cur = cur / my_gcd(cur, (lll)k) * k;
        L[k] = cur;
    }

    lll ans = 0;
    for (int i = 1; i <= 31; i++) {
        lll N = 1;
        for (int j = 0; j < i; j++) N *= 4;
        lll Nm2 = N - 2;

        lll c1;
        if (i == 1) {
            c1 = Nm2;
        } else {
            c1 = (L[i] <= Nm2) ? Nm2 / L[i] : 0;
        }
        lll c2 = (L[i+1] <= Nm2) ? Nm2 / L[i+1] : 0;

        ans += c1 - c2;
    }

    // Print __int128
    long long result = (long long)ans;
    cout << result << endl;

    return 0;
}
